src/enumerable.cr

```diff
@@ -642,6 +662,31 @@ module Enumerable(T)
     index { |e| e == obj }
   end
 
+  # Returns the index of the first element for which the passed block returns `true`.
+  #
+  # ```
+  # ["Alice", "Bob"].index! { |name| name.size < 4 } # => 1 (Bob's index)
+  # ```
+  #
+  # Raises `Enumerable::NotFoundError` if there are no element for which the block is `true`.
+  def index!(& : T ->) : Int32
+    each_with_index do |e, i|
+      return i if yield e
+    end
+    raise Enumerable::NotFoundError.new
+  end
```

Comment:
This is simply `index { |e| yield e } || raise ...`, because a valid index is always non-nil.

Reply:
That is true but I think I prefer the current way it's written.

If I'm looking at 
```crystal
    each_with_index do |e, i|
      return i if yield e
    end
    raise Enumerable::NotFoundError.new
```
It's much clearer that it's following a pattern from
```crystal
    each_with_index do |e, i|
      return i if yield e
    end
    nil
```
than from
```crystal
    index { |e| yield e } || raise Enumerable::NotFoundError.new
```

---

The pattern is *also* consistent with the pattern from `find!` where you can't do this. So let's not cherry-pick places where this happens to be possible.

---

In addition, as written, it avoids a roundtrip through `nil`.

Reply:
I also prefer the more concise way 🤷

Reply:
I updated to the suggested version but I don't really care. Both sides have good arguments and the code is fine either way, IMO.